pkg/controller/sharding/scheduler.go

```diff
@@ -0,0 +1,215 @@
+/*
+Copyright 2023 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sharding
+
+import (
+	"context"
+	"math/rand"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/kubevela/pkg/util/k8s"
+	"github.com/kubevela/pkg/util/maps"
+	velaruntime "github.com/kubevela/pkg/util/runtime"
+	"github.com/kubevela/pkg/util/singleton"
+	"github.com/kubevela/pkg/util/slices"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	"k8s.io/kubectl/pkg/util/podutils"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/oam-dev/kubevela/pkg/utils/util"
+)
+
+// Scheduler schedule shard-id for object
+type Scheduler interface {
+	Start(context.Context)
+	Schedule(client.Object) bool
+}
+
+var _ Scheduler = (*staticScheduler)(nil)
+
+// NewStaticScheduler create a scheduler that do not make update but only use predefined shards for allocate
+func NewStaticScheduler(shards []string) Scheduler {
+	return &staticScheduler{shards: shards}
+}
+
+type staticScheduler struct {
+	shards []string
+}
+
+// Start .
+func (in *staticScheduler) Start(ctx context.Context) {
+	klog.Infof("staticScheduler started, shards: [%s]", strings.Join(in.shards, ", "))
+}
+
+// Schedule the target object to a random shard
+func (in *staticScheduler) Schedule(o client.Object) bool {
+	if _, scheduled := GetScheduledShardID(o); !scheduled {
+		if len(in.shards) > 0 {
+			// nolint
```

Comment:
for what?

Reply:
rand.Intn will be reported as an issue in golang-ci-lint because it is pseudo random number.